oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/OAuth2ClientAuthorizationProvider.java

```diff
@@ -0,0 +1,27 @@
+/*
+ * Copyright 2002-2019 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.springframework.security.oauth2.client;
+
+/**
+ *
+ * @author Joe Grandja
+ * @since 5.2
+ */
+public interface OAuth2ClientAuthorizationProvider<T extends AbstractOAuth2ClientAuthorizationContext> {
```

Comment:
For reactive will need to return a different type `Mono<OAuth2AuthorizedClient>` also. Nowhere else do we try to combine reactive and not reactive, so I don't view this as a good reason for generics.

Similarly, we don't generalize service APIs but instead the types being passed in. Looking at Spring there are variants of `HandlerMethodArgumentResolver` for servlet, messaging, and reactive. I think the same will apply here.

Reply:
> For reactive will need to return a different type Mono<OAuth2AuthorizedClient>

Yes, I'm aware of that but this discussion is about the input parameter which can be reused. As an example, both `WebClientReactiveAuthorizationCodeTokenResponseClient` and `DefaultAuthorizationCodeTokenResponseClient` use the same input `OAuth2AuthorizationCodeGrantRequest`. The same pattern applies here.

If you still believe we should get rid of the generic type than are you suggesting the interface to be changed as follows?

`OAuth2AuthorizedClient authorize(AbstractOAuth2ClientAuthorizationContext authorizationContext);`

Reply:
> Yes, I'm aware of that but this discussion is about the input parameter which can be reused. 

For reactive the output cannot be reused, so we would never use the reactive generic type for the input.

This really doesn't align with code else where and I think just complicates things. Rather than having a generic API I'd like it to be focused on servlet, messaging, or reactive. We won't be able to reuse the implementation and I don't see us ever having a class consume the interface in a way that is not specific to servlet, reactive, or messaging.

> If you still believe we should get rid of the generic type than are you suggesting the interface to be changed as follows?
> 
> OAuth2AuthorizedClient authorize(AbstractOAuth2ClientAuthorizationContext authorizationContext);

I think it would be changed to the following unless you can come up with scenarios where a class would be able to consume the interface with AbstractOAuth2ClientAuthorizationContext as the context vs a concrete subclass.

```
OAuth2AuthorizedClient authorize(ServletOAuth2ClientAuthorizationContext authorizationContext)
```